src/ray/object_manager/pull_manager.cc

```diff
@@ -96,10 +98,21 @@ void PullManager::TryToMakeObjectLocal(const ObjectID &object_id) {
     return;
   }
 
+  // New object locations were found, so begin trying to pull from a
+  // client. This will be called every time a new client location
+  // appears.
+  bool did_pull = PullFromRandomLocation(object_id);
```

Comment:
This is crucial change. We should pull first and try restoring if we cannot pull from anywhere. This will improve the load balancing of pull requests (and also the RPC to the node that spilled the object will decrease).

Reply:
Makes sense, can you add a comment to explain why we must pull first? I think this is also to make sure that we restore properly if it's on a remote disk right? Otherwise, we'll just keep trying to send the restore RPC and never pull the copy that was restored on the remote node.